### ISSUE TITLE
Ubuntu 20.04 support

### ISSUE THREAD

**Issue by icedevml** [tklengyel/drakvuf#866]
It seems like Xen hypervisor built from sources can not be built under Ubuntu 20.04. They've enabled `fcf-protection` by default, which conflicts with Xen's compilation settings.

I've tried with:
```
sed 's/CFLAGS += -mindirect-branch=/CFLAGS += -fcf-protection=none -mindirect-branch=/' -i xen/arch/x86/Rules.mk
```

but then it does compile&install properly, anyway instead the hypervisor is immediately crashing on boot.

Is there anything we could do about this? I don't know what are the particular issues preventing Xen from working.

**Comment by Dos98** [comment]
I asked xen-devel abouut it. https://lists.xenproject.org/archives/html/xen-devel/2020-04/msg01735.html

**Comment by icedevml** [comment]
:thinking: Though I'm doing exactly that - appending `-fcf-protection=none` but it still crashes.

**Comment by Dos98** [comment]
Yes!

**Comment by tklengyel** [comment]
> Is there anything we could do about this?

Install gcc-8 and use that for compiling.

**Comment by tklengyel** [comment]
I've verified that with gcc-8 it compiles and boots and works just fine with Ubuntu 20.04

**Comment by tklengyel** [comment]
@Dos98 Please email back to Andrew on xen-devel the info that with that option it compiles but still crashes.

**Comment by icedevml** [comment]
Btw. how to tell the build system that it should use `gcc-8`? I've passed `CC=gcc-8` and `CXX=g++-8` to `./configure` but the hypervisor part is still compiled using `gcc` command. There are some `$(CC)` references in `xen/xen` Rules.mk but seems liek there is something happening in between.

Right now I've solved this in a brutal way by changing `/usr/bin/gcc` symlink to point to `/usr/bin/gcc-8`, but perhaps there is some better alternative.

**Comment by icedevml** [comment]
Hm, but the above doesn't help anyway, still crashing.

**Comment by tklengyel** [comment]
> Btw. how to tell the build system that it should use gcc-8?

I use update-alternatives for when there are multiple versions of tools available. See https://askubuntu.com/questions/26498/how-to-choose-the-default-gcc-and-g-version

**Comment by icedevml** [comment]
Gonna retry one more time tomorrow, but for now it looks like this gcc-8 build doesn't work for me anyway :(

**Comment by tklengyel** [comment]
Odd, I have it building and working fine with gcc-8

**Comment by icedevml** [comment]
Tested one more time on clean Ubuntu 20.04, with gcc-8 it's not required to patch anything but the resulting Xen is crashing right after "loading initial Xen ramdisk". Giving up for the moment, maybe Xen developers would find a solution soon.

**Comment by icedevml** [comment]
related https://bugs.launchpad.net/ubuntu/+source/gcc-9/+bug/1863260

**Comment by icedevml** [comment]
This allows to build bootable Xen with default Ubuntu 20.04 gcc version:

```
sed 's/CFLAGS += -mindirect-branch=/CFLAGS += -fcf-protection=none -mindirect-branch=/' -i xen/arch/x86/Rules.mk
sed 's/CFLAGS += -Werror/CFLAGS += -Werror -fcf-protection=none/' -i xen/arch/x86/boot/build32.mk
```

**Comment by icedevml** [comment]
package compatibility issues resolved by #868 #869 

known workarounds:
* using `gcc-7` to compile Xen (`ln -s /usr/bin/gcc-7 /usr/bin/gcc`, same for g++)
* using the default `gcc-9` with the above patches to `Rules.mk` and `build32.mk`
